cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -327,6 +345,18 @@ private void handleProgramEvent(ProgramRunId programRunId, ProgramRunStatus prog
     }
   }
 
+  /**
+   * write to heart beat dataset if the recordedRunRecord is not null
+   */
+  private void writeToHeartBeatDataset(@Nullable RunRecordMeta recordedRunRecord,
+                                       long timestampInSeconds, DatasetContext datasetContext) {
+    if (recordedRunRecord != null) {
+      ProgramHeartbeatDataset heartbeatDataset =
+        ProgramHeartbeatDataset.getOrCreate(datasetContext, datasetFramework, cConf);
```

Comment:
The dataset is probably cached, but I think it's still better to call this once and pass around the dataset, similar to what is done for the AppMetadataStore